firefly-api/src/write_node_client.rs

```diff
@@ -0,0 +1,228 @@
+            return Some((sig, datetime, csv_values, cost));
+        })
+        .collect()
+}
+
+/// Processes a vector of tuples containing signatures, timestamps, and CSV values.
+/// Filters out duplicate entries based on signatures, sorts by timestamp,
+/// and returns only the CSV values portion of each tuple.
+///
+/// # Arguments
+/// * `tuples` - Vector of tuples containing (signature, timestamp, csv_values)
+///
+/// # Returns
+/// Vector of tuples with timestamp as Datetime and CSV value vectors, sorted by timestamp with duplicates removed
+fn process_tuples(
+    tuples: Vec<(String, DateTime<Utc>, Vec<String>, u64)>,
+) -> Vec<(String, DateTime<Utc>, Vec<String>, u64)> {
+    let mut seen_sigs = HashSet::new();
+    let mut unique_tuples: Vec<_> = tuples
+        .into_iter()
+        .filter(|(sig, _, _, _)| seen_sigs.insert(sig.to_string()))
+        .collect();
+    unique_tuples.sort_by_key(|(_, datetime, _, _)| *datetime);
+    unique_tuples
+}
+
+#[derive(Debug, Clone)]
+pub struct BlocksClient {
+    http_client: HttpClient,
+    block_node_url: String,
+}
+
+impl BlocksClient {
+    pub fn new(block_node_url: &str) -> Self {
+        let http_client = HttpClient::new();
+        Self {
+            http_client,
+            block_node_url: block_node_url.to_string(),
+        }
+    }
+
+    fn api_url(&self, method: &str) -> String {
+        format!("{}/api/{}", self.block_node_url, method)
+    }
+
+    async fn api_get(
+        &self,
+        method: &str,
+        path_params: Option<&str>,
+        args: Option<&str>,
+    ) -> Result<Value, anyhow::Error> {
+        let url = self.api_url(method);
+        let url = match path_params {
+            Some(params) => format!("{}/{}", url, params),
+            None => url,
+        };
+        let url = match args {
+            Some(args) => format!("{}?{}", url, args),
+            None => url,
+        };
+        let response = self
+            .http_client
+            .get(&url)
+            .header("Content-Type", "text/plain")
+            .send()
+            .await?;
+
+        // Ensure the request was successful
+        if response.status().is_success() {
+            // Parse the JSON response
+            let json_data: Value = response
+                .json()
+                .await
+                .context("failed to parse response as JSON")?;
+            Ok(json_data)
+        } else {
+            Err(anyhow!(
+                "HTTP request failed with status: {}",
+                response.status()
+            ))
+        }
+    }
+
+    async fn first_block_hash(&self) -> Result<String, anyhow::Error> {
+        let response = self.api_get("blocks", None, None).await?;
+
+        Ok(response
+            .as_array()
+            .and_then(|blocks| blocks.get(0))
+            .and_then(|block| block["blockHash"].as_str())
+            .map(|s| s.to_string())
+            .ok_or_else(|| anyhow!("Failed to extract block hash value."))?)
```

Comment:
try to use context by default. it will chain error messages by default
```suggestion
        response
            .as_array()
            .and_then(|blocks| blocks.get(0))
            .and_then(|block| block["blockHash"].as_str())
            .map(|s| s.to_string())
            .context("Failed to extract block hash value")
```

Reply:
ok